src/groth16/prover.rs

```diff
@@ -175,6 +175,306 @@ where
     create_proof::<E, C, P>(circuit, params, r, s)
 }
 
+pub fn create_random_proof_batch<E, C, R, P: ParameterSource<E>>(
+    circuits: Vec<C>,
+    params: P,
+    rng: &mut R,
+) -> Result<Vec<Proof<E>>, SynthesisError>
+where
+    E: Engine,
+    C: Circuit<E> + Send,
+    R: RngCore,
+{
+    let r_s = (0..circuits.len()).map(|_| E::Fr::random(rng)).collect();
+    let s_s = (0..circuits.len()).map(|_| E::Fr::random(rng)).collect();
+
+    create_proof_batch::<E, C, P>(circuits, params, r_s, s_s)
+}
+
+pub fn create_proof_batch<E, C, P: ParameterSource<E>>(
+    circuits: Vec<C>,
+    mut params: P,
+    r_s: Vec<E::Fr>,
+    s_s: Vec<E::Fr>,
+) -> Result<Vec<Proof<E>>, SynthesisError>
+where
+    E: Engine,
+    C: Circuit<E> + Send,
+{
+    // TODO: make sure all circuits have the same input len
+
+    let mut provers = circuits
+        .into_par_iter()
+        .map(|circuit| -> Result<_, SynthesisError> {
+            let mut prover = ProvingAssignment {
+                a_aux_density: DensityTracker::new(),
+                b_input_density: DensityTracker::new(),
+                b_aux_density: DensityTracker::new(),
+                a: vec![],
+                b: vec![],
+                c: vec![],
+                input_assignment: vec![],
+                aux_assignment: vec![],
+            };
+
+            prover.alloc_input(|| "", || Ok(E::Fr::one()))?;
+
+            circuit.synthesize(&mut prover)?;
+
+            for i in 0..prover.input_assignment.len() {
+                prover.enforce(|| "", |lc| lc + Variable(Index::Input(i)), |lc| lc, |lc| lc);
+            }
+
+            Ok(prover)
+        })
+        .collect::<Result<Vec<_>, _>>()?;
+
+    let worker = Worker::new();
+    let input_len = provers[0].input_assignment.len();
+    let vk = params.get_vk(input_len)?;
+    let n = provers[0].a.len();
```

Comment:
Does that mean all circuits have equal number of constraints?

Reply:
yes, this method is specifically designed to process `n` circuits that are identical, just with different values